src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -0,0 +1,258 @@
+
+            var destinationIndex = encodedLength - 4;
+            var sourceIndex = dataLength - leftover;
+            int result = 0;
+
+            ref byte bufferBytes = ref buffer.DangerousGetPinnableReference();
+            // encode last pack to avoid conditional in the main loop
+            if (leftover != 0)
+            {
+                if (leftover == 1)
+                {
+                    result = EncodeAndPadTwo(ref Unsafe.Add(ref bufferBytes, sourceIndex));
+                    Unsafe.WriteUnaligned(ref Unsafe.Add(ref bufferBytes, destinationIndex), result);
+                    destinationIndex -= 4;
+                }
+                else
+                {
+                    result = EncodeAndPadOne(ref Unsafe.Add(ref bufferBytes, sourceIndex));
+                    Unsafe.WriteUnaligned(ref Unsafe.Add(ref bufferBytes, destinationIndex), result);
+                    destinationIndex -= 4;
+                }
+            }
+
+            sourceIndex -= 3;
+            while (sourceIndex >= 0)
+            {
+                result = Encode(ref Unsafe.Add(ref bufferBytes, sourceIndex));
+                Unsafe.WriteUnaligned(ref Unsafe.Add(ref bufferBytes, destinationIndex), result);
+                destinationIndex -= 4;
+                sourceIndex -= 3;
+            }
+
+            written = encodedLength;
+            return OperationStatus.Done;
+
+            FalseExit:
+            written = 0;
+            return OperationStatus.DestinationTooSmall;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Encode(byte b0, byte b1, byte b2, out byte r0, out byte r1, out byte r2, out byte r3)
+        {
+            int i0 = b0 >> 2;
+            r0 = s_encodingMap[i0];
+
+            int i1 = (b0 & 0x3) << 4 | (b1 >> 4);
+            r1 = s_encodingMap[i1];
+
+            int i2 = (b1 & 0xF) << 2 | (b2 >> 6);
+            r2 = s_encodingMap[i2];
+
+            int i3 = b2 & 0x3F;
+            r3 = s_encodingMap[i3];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Encode(byte b0, byte b1, byte b2, out byte r0, out byte r1, out byte r2)
+        {
+            int i0 = b0 >> 2;
+            r0 = s_encodingMap[i0];
+
+            int i1 = (b0 & 0x3) << 4 | (b1 >> 4);
+            r1 = s_encodingMap[i1];
+
+            int i2 = (b1 & 0xF) << 2 | (b2 >> 6);
+            r2 = s_encodingMap[i2];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Encode(byte b0, byte b1, out byte r0, out byte r1)
+        {
+            int i0 = b0 >> 2;
+            r0 = s_encodingMap[i0];
+
+            int i1 = (b0 & 0x3) << 4 | (b1 >> 4);
+            r1 = s_encodingMap[i1];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int Encode(ref byte threeBytes)
+        {
```

Comment:
You should be able to use same set of tricks here as you are using for Decode: cache the map in a local upfront, avoid bound checks for it and save some binary ops by operating on the whole byte. Something like:

```
private static int Encode(ref byte threeBytes, ref byte encodingMap)
{
    int i = (threeBytes << 16) | (Unsafe.Add(ref threeBytes, 1) << 8) | Unsafe.Add(ref threeBytes, 2);

    int i0 = Unsafe.Add(ref encodingMap, i >> 18)
    int i1 = Unsafe.Add(ref encodingMap, (i >> 12) & 0x3F);
    int i2 = Unsafe.Add(ref encodingMap, (i >> 6) & 0x3F);
    int i3 = Unsafe.Add(ref encodingMap, i & 0x3F);

    return i0 | (i1 << 8) | (i2 << 16) | (i3 << 24);
}
```